### PR TITLE
Add Brave suggested sites

### PR DESCRIPTION
**Pull request by bbondy**
Fix https://github.com/brave/brave-browser/issues/9302

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/omnibox/browser/suggested_sites_provider.cc [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+// static
+ACMatchClassifications SuggestedSitesProvider::StylesForSingleMatch(
+    const std::string &input_text,
+    const std::string &site) {
+  ACMatchClassifications styles;
+  size_t foundPos = site.find(input_text);
+  if (std::string::npos == foundPos) {
+    styles.push_back(ACMatchClassification(0, ACMatchClassification::NONE));
+    return styles;
```
Comment (emerick) [inline components/omnibox/browser/suggested_sites_provider.cc:78]: Not a big deal to me, but wasn't sure why we're doing a direct return here versus at the end of the function. Seems like that's how we handle the other two clauses of the `if`.

File: components/omnibox/browser/topsites_provider_unittest.cc [hunk 2]
```diff
@@ -19,7 +20,7 @@ class TopSitesProviderTest : public testing::Test {
                             metrics::OmniboxEventProto::OTHER,
                             classifier_);
     return input;
- }
+  }
```
Comment (emerick) [inline components/omnibox/browser/topsites_provider_unittest.cc:23]: Nit: whitespace here seems unintentional.

File: components/omnibox/browser/suggested_sites_match.cc [hunk 3]
```diff
@@ -0,0 +1,21 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/omnibox/browser/suggested_sites_match.h"
+
+// This is the provider for Brave Suggested Sites
+SuggestedSitesMatch::SuggestedSitesMatch(GURL destination_url,
```
[the remark this answers is not in the document]
Comment (bbondy) [reply components/omnibox/browser/suggested_sites_match.cc:9]: yep, will update.

File: components/omnibox/browser/suggested_sites_provider.cc [hunk 4]
```diff
@@ -0,0 +1,109 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (bbondy) [reply components/omnibox/browser/suggested_sites_provider.cc:1]: Off by 1 error ;)

File: components/omnibox/browser/suggested_sites_provider.cc [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    // really want these suggestions. Example don't suggest bitcoin and
+    // litecoin for just a coin search.
+    if (foundPos == 0) {
+      ACMatchClassifications styles =
+          StylesForSingleMatch(input_text,
+              base::UTF16ToASCII(match.display));
+      AddMatch(base::ASCIIToUTF16(current_site), match, styles);
+      if (match.allow_default &&
+          current_site.length() == input_text.length()) {
+        size_t last_index = matches_.size() - 1;
```
[the remark this answers is not in the document]
Comment (bbondy) [reply components/omnibox/browser/suggested_sites_provider.cc:57]: It is due to the AddMath call a couple lines before. But I added a comment there now 👍

File: components/omnibox/browser/suggested_sites_provider.cc [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        matches_[last_index].SetAllowedToBeDefault(input);
+        // As from autocomplete_provider.h:
+        // Search Primary Provider (what you typed) | 1300
+        matches_[last_index].relevance = 1301;
```
[the remark this answers is not in the document]
Comment (bbondy) [reply components/omnibox/browser/suggested_sites_provider.cc:61]: yep, there's a comment in chromium source that the weird relevance numbers should be refactored.

File: components/omnibox/browser/topsites_provider_unittest.cc (hunk 2 above)
Comment (bbondy) [reply components/omnibox/browser/topsites_provider_unittest.cc:23]: It's actually an intentional lint correction, only changed files are tested in a PR so this file was ignored before.  👍

File: components/omnibox/browser/suggested_sites_provider.cc (hunk 1 above)
Comment (bbondy) [reply components/omnibox/browser/suggested_sites_provider.cc:80]: you're right, fixed.

File: components/omnibox/browser/suggested_sites_match.cc [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    GURL stripped_destination_url, base::string16 display, bool allow_default) :
+  destination_url(destination_url),
```
[the remark this answers is not in the document]
Comment (bbondy) [reply components/omnibox/browser/suggested_sites_match.cc]: updated

**Comment by bbondy** [comment]
per a good suggestion from @diracdeltas, I also added in the ref code for the displayed params

**Comment by Shaftw** [comment]
Thank fuck I stayed away from Brave

**Comment by Alex-CodeLab** [comment]
scammers gonna scam

**Comment by Kukks** [comment]
This is some of the shadiest shit I have seen being done to a browser. Thank fuck I chose to use Firefox over brave.

**Comment by diracdeltas** [comment]
> Thanks for the feedback guys.
> 
> I don't claim we're free from misjudgments from time to time but we 1) work in the open and 2) listen to and value user feedback.
> 
> Could I dig deeper to get more feedback? I'd like to know if the problem is the suggestions themselves (I don't think so) or more precisely that we made a misjudgement with a default match to one of those suggestions.
> 
> This is the block of code that I think rubs people the wrong way
> https://github.com/brave/brave-core/blob/master/components/omnibox/browser/suggested_sites_provider.cc#L64
> 
> Or more precisely we could set `allow_default_` to `false` for all entries in the last param here https://github.com/brave/brave-core/blob/master/components/omnibox/browser/suggested_sites_provider_data.cc

sorry for top-posting, idk why but github isn't letting me reply to a threaded comment.

i have a few suggestions:

1. these refcodes look high-entropy, but IIUC in reality they're the same as having Brave in the user-agent (which we are doing soon anyway?) or referer. would these sites be ok with us using the `brave=true` parameter instead of `refcode=...`? at least that would make it more transparent to the user what is going on.
2. regardless i think the behavior upon hitting enter in the urlbar, which is what people do the vast majority of the time to initiate a navigation, should NOT inject anything. right now if you type binance.com and hit enter, it navigates to binance.com/refcode=...

**Comment by bbondy** [comment]
@diracdeltas I commented on both of those points this morning in the commit: https://github.com/brave/brave-core/commit/e8fdde70a3ac2c25e6ccc567bc93c6d4059c2d72#commitcomment-39750337

**Pull request merged**